DOCSP-29928 Serialization for Dart (#3235)
mongodb/docs-realm: source/sdk/flutter/realm-database/serialization.txt

.. _flutter-serialization:

===========================
Serialization - Flutter SDK
===========================

.. facet::
  :name: genre
  :values: reference

.. meta::
  :description: Serialize objects to and from EJSON with a built-in feature for Atlas Device SDK for Flutter.

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

The Atlas Device SDK for Flutter supports serialization and deserialization of
:manual:`EJSON </reference/mongodb-extended-json>` to and from static Realm objects.
Serialization allows for easier integration with MongoDB APIs.

Supported Data Types for Serialization
--------------------------------------

> The following :flutter-sdk:`defined enum values <realm/EJsonType.html>` 
> can be serialized and deserialized:

Reviewer: ```suggestion
The Flutter SDK supports serializing the following data types:
```
(sorry again; I know I'm the one to mention the enum values link, but I'm not sure how helpful it actually is 🙈 )

- Array
- Binary
- Boolean
- Date

> - Decimal128

Reviewer: ```suggestion
- Document
- Decimal128
```
Document is listed as a supported enum value

- Double
- Int32
- Int64
- MaxKey
- MinKey
- ObjectID
- String
- Symbol
- Null
- Undefined

> - :ref:`RealmValue <flutter-realm-value>`

Reviewer: ```suggestion
```
I know I was the one to ask about this (and I might be confusing myself at this point), but I think we're only listing the BSON types here, not the Realm-specific types

.. _flutter-serialization:

===========================
Serialization - Flutter SDK
===========================

.. facet::
  :name: genre
  :values: reference

.. meta::
  :description: Serialize objects to and from EJSON with a built-in feature for Atlas Device SDK for Flutter.

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

The Atlas Device SDK for Flutter supports serialization and deserialization of
:manual:`Extended JSON (ESJON) </reference/mongodb-extended-json>` to and from static Realm objects.
Serialization allows for easier integration with MongoDB APIs.

Supported Data Types for Serialization
--------------------------------------

The Flutter SDK supports serializing the following data types:

- Array
- Binary
- Boolean
- Date
- Document
- Decimal128
- Double
- Int32
- Int64
- MaxKey
- MinKey
- ObjectID
- String
- Symbol
- Null

> - Undefined

Reviewer: Agreed. Can we make this a table, @lindseymoore? Alternately, we could make it a table of:  
`Dart/Realm type | Serialized type | Example output`

.. list-table::
   :header-rows: 1
   :widths: 30 70

   * - Realm Type
     - Serializes To

   * - DateTime
     - Date

       ``DateTime birthDate = DateTime.utc(2024, 4, 10)`` serializes to ``birthDate: {$date: {$numberLong: 1712707200000}}``

   * - RealmList
     - Array

       ``List<String> listOfStrings = [food, water]`` serializes to ``listOfStrings: [food, water]``

   * - RealmMap
     - Array

       ``Map<String, int> mapOfMixedAnyValues = {'first': 123 , 'second': 567}`` serializes to ``mapOfValues: {first: {$numberInt: 123}, second: {$numberInt: 567}}``

   * - RealmSet
     - Array

       ``Set<int> setOfInts = {0, 1, 2, 3}`` serializes to ``setOfInts: [{$numberInt: 0}, {$numberInt: 1}, {$numberInt: 2}, {$numberInt: 3}]``

   * - ObjectId
     - ObjectId

       ``ObjectId id = ObjectId()`` serializes to ``{id: {$oid: 666a6fd54978af08e54a8d52}``

   * - UUID
     - Binary

       ``Uuid myId = Uuid.v4()`` serializes to ``myId: {$binary: {base64: 6TvsMWxDRWa1jSC6gxiM3A==, subType: 04}}``

   * - Uint8List
     - Binary

       ``Uint8List aBinaryProperty = Uint8List.fromList([1, 2])`` serializes to ``aBinaryProperty: {$binary: {base64: AQI=, subType: 00}}``

>    * - Embedded Object

Reviewer: I think we should be okay with just Object here. To show that any Object <-> Document

:flutter-sdk:`toEjson() <realm/toEJson.html>` method in the following two ways: 

.. io-code-block::
    :copyable: true
    
    .. input:: /examples/generated/flutter/serialization_test.snippet.serialize.dart
        :language: dart

    .. output:: 

        {
            type: Jumping Spider, 
            numberOfLegs: {$numberInt: 8}, 
            birthDate: {$date: {$numberLong: 1712707200000}}, 
            price: null
        }

Deserialize from EJSON
~~~~~~~~~~~~~~~~~~~~~~

Deserialize from EJSON using the :flutter-sdk:`fromEjson() <realm/fromEJson.html>`
method. The method takes EJSON for a specified object type as input and outputs 
a deserialized instance of the specified object type. 

The following example deserializes ``serializeByParam`` from the previous example:

.. literalinclude:: /examples/generated/flutter/serialization_test.snippet.deserialize.dart
    :language: dart

Serialize Non-Realm Objects
---------------------------

For non-Realm classes, you can use the ``@ejson`` annotation on the class constructor
to generate a decoder and encoder:

.. literalinclude:: /examples/generated/flutter/person.snippet.ejson-annotation.dart

>     :language: dart

Reviewer: ```suggestion
    :language: dart
    :emphasize-lines: 9
```
to highlight the @ejson annotation line

.. _flutter-serialization:

===========================
Serialization - Flutter SDK
===========================

.. facet::
  :name: genre
  :values: reference

.. meta::
  :description: Serialize objects to and from EJSON with a built-in feature for Atlas Device SDK for Flutter.

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

The Atlas Device SDK for Flutter supports serialization and deserialization of
:manual:`Extended JSON (ESJON) </reference/mongodb-extended-json>` to and from static Realm objects.

> Serialization allows for easier integration with MongoDB APIs.

Reviewer: ```suggestion
The Flutter SDK can use this full-document EJSON encoding to communicate 
directly with MongoDB Atlas, such as through the following APIs:

- App Services :ref:`Function calls <flutter-call-function>`
- Credentials with :ref:`custom function authentication <flutter-login-custom-function>`
- User profile and :ref:`custom user data <flutter-custom-user-data>`
```
You'll also need to check these linked pages and make sure there aren't any doc impacts (e.g. the Call a Function page [instructions](https://www.mongodb.com/docs/atlas/device-sdks/sdk/flutter/app-services/call-function/#call-a-function:~:text=To%20include%20objects,native%20Dart%20object.) need to be updated with this new ESJON method)

fyi - this list will eventually include the [enhanced Query MongoDB functionality](https://jira.mongodb.org/browse/DOCSP-37312), which hasn't been released yet

Author: Ticket for this work: https://jira.mongodb.org/browse/DOCSP-41296

Reviewer: I think the suggested edits here are still valid, even if the linked pages will be updated in separate tickets (let's make sure we have tickets tracking the custom function auth updates too. doesn't look like custom user data has any impacts)

Author: I see the function on this page converts to json rather than ejson. Is converting to ejson the more updated way; will it no longer work with json?
I'll add the function auth update to the previous ticket!

Reviewer: ooh, good catch! 🕵️ looks like that call uses `jsonEncode()` and `jsonDecode()` internally, so we should probably still point users to serialize object args that way. Although now I'm a little confused by the last sentence... ("The Function returns a dynamic value containing MongoDB Extended JSON (EJSON) deserialized to a native Dart object.") 

Either way, I think we're good to leave this as-is. Thanks for digging into this and confirming!

Supported Data Types for Serialization
--------------------------------------

Flutter SDK supports the following data types:

.. list-table::
   :header-rows: 1
   :widths: 30 30 40

>    * - Realm Data Type

Reviewer: these aren't technically all Realm Data Types; they're Dart + Realm-specific types. It might make sense to only include the realm-specific ones since the Dart types serialize to their ejson counterparts (bool > bool, string > string)

> The Flutter SDK supports serialization of all :ref:`supported data types <flutter-data-types>`. 

Reviewer: ```suggestion
The Flutter SDK currently supports serialization of the following :ref:`supported data types <flutter-data-types>`: 

- All Dart-language data types
- All Realm-specific data types _except_ Decimal128 and RealmValue
```
this lets us move these 2 realm-specific types from the list of unsupported BSON types below

   * - RealmSet
     - Array

       ``Set<int> setOfInts = {0, 1, 2, 3}`` serializes to ``setOfInts: [{$numberInt: 0}, {$numberInt: 1}, {$numberInt: 2}, {$numberInt: 3}]``

   * - ObjectId
     - ObjectId

       ``ObjectId id = ObjectId()`` serializes to ``{id: {$oid: 666a6fd54978af08e54a8d52}``

   * - UUID
     - Binary

       ``Uuid myId = Uuid.v4()`` serializes to ``myId: {$binary: {base64: 6TvsMWxDRWa1jSC6gxiM3A==, subType: 04}}``

   * - Uint8List
     - Binary

       ``Uint8List aBinaryProperty = Uint8List.fromList([1, 2])`` serializes to ``aBinaryProperty: {$binary: {base64: AQI=, subType: 00}}``

   * - Object
     - Document
     
       ``Address address = Address("500 Dean Street", "Brooklyn", "NY", "USA")`` serializes to ``address: {street: 500 Dean Street, city: Brooklyn, state: NY, country: USA}``
    
For more information on the serialization of non-Realm specific types,
see :manual:`BSON Data Types and Associated Representations </reference/mongodb-extended-json/#bson-data-types-and-associated-representations>`.

> The SDK does not currently support the following types: RealmValue, Decimal128, Code, 
> CodeWScope, DBPointer, DBRef, Regular Expression, and Timestamp. 

Reviewer: ```suggestion
Flutter SDK serialization does not currently support the following BSON types: Code, 
CodeWScope, DBPointer, DBRef, Regular Expression, and Timestamp. 
```

CodeWScope, DBPointer, DBRef, Regular Expression, and Timestamp. 

Object Model for Serialization
------------------------------

Full-document encoding enables you to serialize and deserialize user-defined 
classes. You can create your object model as you normally would. The RealmObject class
model created by your ``part`` declaration provides the necessary encoder 
and decoder methods for serialization and deserialization. The following ``Pet``
object model will be used in the following examples:

.. literalinclude:: /examples/generated/flutter/pet.snippet.serialize-object-model.dart
    :language: dart

Serialization
-------------

For objects based on ``RealmObject`` classes, you can serialize to EJSON using the 
:flutter-sdk:`toEjson() <realm/toEJson.html>` method in the following two ways: 

.. io-code-block::
    :copyable: true
    
    .. input:: /examples/generated/flutter/serialization_test.snippet.serialize.dart
        :language: dart

    .. output:: 

        {
            type: Jumping Spider, 
            numberOfLegs: {$numberInt: 8}, 
            birthDate: {$date: {$numberLong: 1712707200000}}, 
            price: null
        }

Deserialization

> ---------------

Author: Keeping as a regular heading, but let me know.

.. _flutter-serialization:

===========================
Serialization - Flutter SDK
===========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

The Atlas Device SDK for Flutter supports serialization and deserialization of
:manual:`EJSON </reference/mongodb-extended-json>` to and from static Realm objects.
Serialization allows for easier integration with MongoDB APIs.

Supported Data Types for Serialization
--------------------------------------

The following data types can be serialized and deserialized:

- Array
- Binary
- Boolean
- Date
- Decimal128
- Double
- Int32
- Int64
- MaxKey
- MinKey
- ObjectID
- String
- Symbol
- Null
- Undefined

> 

Reviewer 2: Should we mention the types we don't support? (yet?), like `Code`, `CodeWScope`, `DBPointer`, `DBRef`, `Regular Expression`, and `Timestamp`(which is nothing like you would expect).

Not sure myself. What does other SDKs do?

Author: I think it would be helpful to include!

For more information and examples on the serialization for each of these types,
see :manual:`BSON Data Types and Associated Representations </reference/mongodb-extended-json/#bson-data-types-and-associated-representations>`.

Object Model for Serialization
------------------------------

Full-document encoding enables you to serialize and deserialize user-defined 
classes. You can create your object model as you normally would. The RealmObject class
model created by your ``part`` declaration will provide the neccessary encoder 
and decoder methods for serialization and deserialization. The following ``Pet``
object model will be used in the following examples:

.. literalinclude:: /examples/generated/flutter/pet.snippet.serialize-object-model.dart
    :language: dart

Serialization
-------------

For objects in a Realm, you can serialize to EJSON using the ``toEjson()`` method. 

.. literalinclude:: /examples/generated/flutter/serialization_test.snippet.serialize.dart
    :language: dart

The outputted EJSON looks like the following:

.. code-block:: EJSON

>     {
>     type: Jumping Spider, 
>     numberOfLegs: {$numberInt: 8}, 
>     birthDate: {$date: {$numberLong: 1712707200000}}, 
>     price: null
>     }

Reviewer 2: ```suggestion
    {
      type: Jumping Spider, 
      numberOfLegs: {$numberInt: 8}, 
      birthDate: {$date: {$numberLong: 1712707200000}}, 
      price: null
    }
```
Should we indent a bit?

.. _flutter-serialization:

===========================
Serialization - Flutter SDK
===========================

.. facet::
  :name: genre
  :values: reference

.. meta::
  :description: Serialize objects to and from EJSON with a built-in feature for Atlas Device SDK for Flutter.

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

> The Atlas Device SDK for Flutter supports serialization and deserialization of

Author: Added information about the @ejson annotation and mention of the API ref for register based on input from Kasper.

:manual:`EJSON </reference/mongodb-extended-json>` to and from static Realm objects.
Serialization allows for easier integration with MongoDB APIs.

Supported Data Types for Serialization
--------------------------------------

> The following data types can be serialized and deserialized:

Author: Undefined is a created class that serializes to `{'\$undefined': 1}`. Perhaps I can include the serialization as a note?
RealmValue is supported. Added to the list.
